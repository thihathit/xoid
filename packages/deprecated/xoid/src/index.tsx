export type { Init, Atom, Listener, StateOf, GetState } from '@xoid/engine'
export type { Usable, Enhancer } from './core'

export { subscribe, effect } from '@xoid/engine'
export { create, use } from './core'
