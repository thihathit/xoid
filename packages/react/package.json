{
  "name": "@xoid/react",
  "version": "1.0.0-beta.0",
  "main": "./index.js",
  "module": "./index.esm.js",
  "types": "./index.d.ts",
  "sideEffects": false,
  "description": "Framework-agnostic state management library designed for simplicity and scalability",
  "homepage": "https://xoid.dev",
  "repository": {
    "type": "git",
    "url": "https://github.com/onurkerimov/xoid",
    "directory": "./packages/react"
  },
  "license": "MIT",
  "keywords": [
    "framework-agnostic",
    "state",
    "management",
    "state-management",
    "react",
    "redux",
    "recoil",
    "xoid",
    "atom",
    "store",
    "stream",
    "observable"
  ],
  "dependencies": {
    "use-sync-external-store": "^1.1.0"
  },
  "peerDependencies": {
    "xoid": ">=1.0.0-beta.0",
    "react": "*"
  }
}