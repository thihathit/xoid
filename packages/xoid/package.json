{
  "name": "xoid",
  "version": "1.0.0-beta.0",
  "main": "./index.js",
  "module": "./index.esm.js",
  "types": "./index.d.ts",
  "sideEffects": false,
  "description": "Framework-agnostic state management library designed for simplicity and scalability",
  "homepage": "https://xoid.dev",
  "repository": {
    "type": "git",
    "url": "https://github.com/onurkerimov/xoid",
    "directory": "./packages/xoid"
  },
  "license": "MIT",
  "keywords": [
    "framework-agnostic",
    "state",
    "management",
    "state-management",
    "react",
    "redux",
    "recoil",
    "xoid",
    "atom",
    "store",
    "stream",
    "observable"
  ]
}