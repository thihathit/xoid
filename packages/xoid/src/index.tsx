import type { Atom, Stream, Init, GetState, Usable } from './types'
import { createSelector, createApi, INTERNAL } from './internal/utils'
import { createEvent, createInternal } from './internal/lite'

export * from './types'

/**
 * Creates an atom with the first argument as the initial state.
 * Second argument can be used to attach more actions to the atom.
 * @see [xoid.dev/docs/api/create](https://xoid.dev/docs/api/create)
 */
export function create<T>(): Stream<T>
export function create<T>(init: Init<T>): Atom<T>
export function create<T>(init: Init<T>, getUsable: null): Atom<T>
export function create<T, U>(init: Init<T>, getUsable?: (atom: Atom<T>) => U): Atom<T> & Usable<U>
export function create<T, U = undefined>(
  init?: Init<T>,
  getUsable?: null | ((atom: Atom<T>) => U)
) {
  const isFunction = typeof init === 'function'
  const initialValue = (isFunction ? undefined : init) as T
  const internal = createInternal(initialValue, (() => (use as any).devtools.send(atom)) as any)
  if (isFunction) internal.get = createSelector(internal, init as (get: GetState) => T)
  const atom = createApi(internal)
  internal.usable = getUsable?.(atom)
  // devtools support
  Object.defineProperty(atom, 'debugValue', {
    set(debugValue: string) {
      internal.debugValue = debugValue
      ;(use as any).devtools.init(atom, initialValue, isFunction)
    },
  })
  return atom
}

/**
 * Gets the "usables" of an atom.
 * @see [xoid.dev/docs/api/use](https://xoid.dev/docs/api/use)
 */
export const use = <T extends any>(atom: Usable<T>): T =>
  (use as any).devtools.wrap((atom as any)[INTERNAL].usable, atom)
;(use as any).symbol = INTERNAL
;(use as any).createEvent = createEvent
;(use as any).devtools = {
  init: <T,>(_atom: T) => void 0,
  send: <T,>(_atom: T) => void 0,
  wrap: <T,>(value: T): T => value,
}
