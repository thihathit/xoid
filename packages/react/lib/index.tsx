import { useSyncExternalStore } from 'use-sync-external-store/shim'
import { useEffect, useLayoutEffect, useRef, useMemo, useDebugValue } from 'react'
import { Atom, create } from 'xoid'
import { createCleanup, createReadable, META, OnCleanup } from '@xoid/engine'

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
const useIsoLayoutEffect = window === undefined ? useEffect : useLayoutEffect

const useConstant = <T extends any>(fn: () => T): T => {
  const ref = useRef<{ c: T }>()
  if (!ref.current) ref.current = { c: fn() }
  return ref.current.c
}

const useCleanup = () => {
  const { onCleanup, cleanupAll } = useConstant(createCleanup)
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useIsoLayoutEffect(() => cleanupAll, [])
  return onCleanup
}

/**
 * Subscribes to a xoid atom inside a React function component.
 * @see [xoid.dev/docs/api-react/useatom](https://xoid.dev/docs/api-react/useatom)
 */
export function useAtom<T>(atom: Atom<T>): T
export function useAtom<T, U>(atom: Atom<T>, selector: (state: T) => U): U
export function useAtom<T, U extends keyof T>(atom: Atom<T>, selector: U): T[U]
export function useAtom<T, U>(atom: Atom<T>, selector?: keyof T | ((state: T) => U)): any {
  const readable = useMemo(() => createReadable(atom as Atom<T>, selector), [atom, selector])
  const result = useSyncExternalStore(
    (readable as any)[META].notifier.subscribe,
    readable,
    readable
  )
  useDebugValue(result)
  return result
}

/**
 * Can be used to create local state inside React components. Similar to `React.useMemo`.
 * @see [xoid.dev/docs/api-react/usesetup](https://xoid.dev/docs/api-react/usesetup)
 */
export function useSetup<T>(setupFn: (deps: Atom<undefined>, onCleanup: OnCleanup) => T): T
export function useSetup<T, P>(setupFn: (deps: Atom<P>, onCleanup: OnCleanup) => T, props: P): T
export function useSetup(setupFn: (deps: any, onCleanup: any) => any, props?: any): any {
  const $deps = useConstant(() => create(() => props))
  // Don't refactor line 51, because it needs to be strictly `() => props`
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useIsoLayoutEffect(() => $deps(() => props), [props])
  const onCleanup = useCleanup()
  return useConstant(() => setupFn($deps, onCleanup))
}
