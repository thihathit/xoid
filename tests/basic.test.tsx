import React, { StrictMode } from 'react'
import ReactDOM from 'react-dom'
import { act, cleanup, fireEvent, render } from '@testing-library/react'
import { create, use, Value } from 'xoid'
import { useAtom } from '@xoid/react'
import { debug } from './testHelpers'

const consoleError = console.error
afterEach(() => {
  cleanup()
  console.error = consoleError
})

it('creates a atom with a primitive value', () => {
  const atom = create(5)
  expect(debug(atom)).toMatchSnapshot()
})

it('creates a atom with a record', () => {
  const atom = create({ alpha: 3, beta: 5 })
  expect(debug(atom)).toMatchSnapshot()
})

it('normalizes nested atoms in a record', () => {
  const atom = create({ alpha: create(3), beta: create(5) })
  expect(debug(atom)).toMatchSnapshot()
})

it('uses the actions in vanilla', async () => {
  const atom = create({ count: 0 }, (atom) => ({
    inc: () => atom.update((state) => ({ count: state.count + 1 })),
  }))
  use(atom).inc()
  expect(debug(atom)).toMatchSnapshot()
})

it('uses the actions in React', async () => {
  const atom = create({ count: 0 }, (atom) => ({
    inc: () => atom.update((state) => ({ count: state.count + 1 })),
  }))

  function Counter() {
    const { count } = useAtom(atom)
    const { inc } = use(atom)
    React.useEffect(inc, [inc])
    return <div>count: {count}</div>
  }

  const { findByText } = render(<Counter />)

  await findByText('count: 1')
})

it('only runs when partial state changes in React', async () => {
  const atom = create({ count: 0, count2: 'constant' }, (atom) => ({
    inc: () => atom.update((state) => ({ ...state, count: state.count + 1 })),
  }))

  let renderCount = 0

  function Counter() {
    const c2 = useAtom(atom.focus((s) => s.count2))
    React.useEffect(use(atom).inc, [])
    renderCount++
    return <div>count: {c2}</div>
  }

  const { findByText } = render(<Counter />)

  await findByText('count: constant')
  expect(renderCount).toBe(1)
})

it('can batch updates', async () => {
  const atom = create(
    {
      count: 0,
    },
    (atom) => ({
      inc: () => atom.update((state) => ({ count: state.count + 1 })),
    })
  )

  function Counter() {
    const { count } = useAtom(atom)
    React.useEffect(() => {
      const { inc } = use(atom)
      ReactDOM.unstable_batchedUpdates(() => {
        inc()
        inc()
      })
    }, [])
    return <div>count: {count}</div>
  }

  const { findByText } = render(<Counter />)

  await findByText('count: 2')
})

it('can update the selector', async () => {
  const atom = create(() => ({
    one: 'one',
    two: 'two',
  }))

  type State = Value<typeof atom>

  function Component({ selector }: any) {
    const value = useAtom(atom.focus(selector))
    return <div>{value}</div>
  }

  const { findByText, rerender } = render(<Component selector={(s: State) => s.one} />)
  await findByText('one')

  rerender(<Component selector={(s: State) => s.two} />)
  await findByText('two')
})

it('ensures parent components subscribe before children', async () => {
  type State = {
    children: { [key: string]: { text: string } }
  }
  type Props = { id: string }
  const atom = create<State>(() => ({
    children: {
      '1': { text: 'child 1' },
      '2': { text: 'child 2' },
    },
  }))

  function changeState() {
    atom.set({
      children: {
        '3': { text: 'child 3' },
      },
    })
  }

  function Child({ id }: Props) {
    // In zustand's tests, the selector in the following line uses optional chaining.
    // In xoid, focus function cannot use optional chaining, so the test is modified.
    const { text } = useAtom(atom.focus((s) => s.children[id]))
    return <div>{text}</div>
  }

  function Parent() {
    const childStates = useAtom(atom.focus((s) => s.children))
    return (
      <>
        <button onClick={changeState}>change state</button>
        {Object.keys(childStates).map((id) => (
          <Child id={id} key={id} />
        ))}
      </>
    )
  }

  const { getByText, findByText } = render(
    <StrictMode>
      <Parent />
    </StrictMode>
  )

  fireEvent.click(getByText('change state'))

  await findByText('child 3')
})

// https://github.com/pmndrs/zustand/issues/84
it('ensures the correct subscriber is removed on unmount', async () => {
  const atom = create({ count: 0 })

  function increment() {
    atom.update(({ count }) => ({ count: count + 1 }))
  }

  function Count() {
    const c = useAtom(atom.focus((s) => s.count))
    return <div>count: {c}</div>
  }

  function CountWithInitialIncrement() {
    React.useLayoutEffect(increment, [])
    return <Count />
  }

  function Component() {
    const [Counter, setCounter] = React.useState(() => CountWithInitialIncrement)
    React.useLayoutEffect(() => {
      setCounter(() => Count)
    }, [])
    return (
      <>
        <Counter />
        <Count />
      </>
    )
  }

  const { findAllByText } = render(<Component />)

  expect((await findAllByText('count: 1')).length).toBe(2)

  act(increment)

  expect((await findAllByText('count: 2')).length).toBe(2)
})

// https://github.com/pmndrs/zustand/issues/86
it('ensures a subscriber is not mistakenly overwritten', async () => {
  const atom = create({ count: 0 })

  function Count1() {
    const c = useAtom(atom.focus((s) => s.count))
    return <div>count1: {c}</div>
  }

  function Count2() {
    const c = useAtom(atom.focus((s) => s.count))
    return <div>count2: {c}</div>
  }

  // Add 1st subscriber.
  const { findAllByText, rerender } = render(<Count1 />)

  // Replace 1st subscriber with another.
  rerender(<Count2 />)

  // Add 2 additional subscribers.
  rerender(
    <>
      <Count2 />
      <Count1 />
      <Count1 />
    </>
  )

  // Call all subscribers
  act(() => atom.set({ count: 1 }))

  expect((await findAllByText('count1: 1')).length).toBe(2)
  expect((await findAllByText('count2: 1')).length).toBe(1)
})
