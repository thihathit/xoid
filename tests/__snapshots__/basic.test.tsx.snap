// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a atom with a primitive value 1`] = `
Object {
  "get": 5,
  "getSerialized": "5",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": 5,
    "watch": [Function],
    Symbol(): Object {
      "get": [Function],
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": undefined,
    },
  },
  "selfSerialized": "{\\"value\\":5}",
  "use": undefined,
}
`;

exports[`creates a atom with a record 1`] = `
Object {
  "get": Object {
    "alpha": 3,
    "beta": 5,
  },
  "getSerialized": "{\\"alpha\\":3,\\"beta\\":5}",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": Object {
      "alpha": 3,
      "beta": 5,
    },
    "watch": [Function],
    Symbol(): Object {
      "get": [Function],
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": undefined,
    },
  },
  "selfSerialized": "{\\"value\\":{\\"alpha\\":3,\\"beta\\":5}}",
  "use": undefined,
}
`;

exports[`normalizes nested atoms in a record 1`] = `
Object {
  "get": Object {
    "alpha": Object {
      "focus": [Function],
      "map": [Function],
      "set": [Function],
      "subscribe": [Function],
      "update": [Function],
      "value": 3,
      "watch": [Function],
      Symbol(): Object {
        "get": [Function],
        "listeners": Set {},
        "set": [Function],
        "subscribe": [Function],
        "usable": undefined,
      },
    },
    "beta": Object {
      "focus": [Function],
      "map": [Function],
      "set": [Function],
      "subscribe": [Function],
      "update": [Function],
      "value": 5,
      "watch": [Function],
      Symbol(): Object {
        "get": [Function],
        "listeners": Set {},
        "set": [Function],
        "subscribe": [Function],
        "usable": undefined,
      },
    },
  },
  "getSerialized": "{\\"alpha\\":{\\"value\\":3},\\"beta\\":{\\"value\\":5}}",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": Object {
      "alpha": Object {
        "focus": [Function],
        "map": [Function],
        "set": [Function],
        "subscribe": [Function],
        "update": [Function],
        "value": 3,
        "watch": [Function],
        Symbol(): Object {
          "get": [Function],
          "listeners": Set {},
          "set": [Function],
          "subscribe": [Function],
          "usable": undefined,
        },
      },
      "beta": Object {
        "focus": [Function],
        "map": [Function],
        "set": [Function],
        "subscribe": [Function],
        "update": [Function],
        "value": 5,
        "watch": [Function],
        Symbol(): Object {
          "get": [Function],
          "listeners": Set {},
          "set": [Function],
          "subscribe": [Function],
          "usable": undefined,
        },
      },
    },
    "watch": [Function],
    Symbol(): Object {
      "get": [Function],
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": undefined,
    },
  },
  "selfSerialized": "{\\"value\\":{\\"alpha\\":{\\"value\\":3},\\"beta\\":{\\"value\\":5}}}",
  "use": undefined,
}
`;

exports[`uses the actions in vanilla 1`] = `
Object {
  "get": Object {
    "count": 1,
  },
  "getSerialized": "{\\"count\\":1}",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": Object {
      "count": 1,
    },
    "watch": [Function],
    Symbol(): Object {
      "get": [Function],
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": Object {
        "inc": [Function],
      },
    },
  },
  "selfSerialized": "{\\"value\\":{\\"count\\":1}}",
  "use": Object {
    "inc": [Function],
  },
}
`;
